source/common/network/proxy_receive_before_connect.h

```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include "envoy/stream_info/filter_state.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * Keep downstream read enabled before connecting upstream. This is expected to be checked for by
+ * any proxy (e.g., tcp_proxy) that can buffer received downstream data to be written upstream once
+ * the upstream connection has been established.
+ *
+ * Any filter in the filter chain can set this to 'true' to request access to data received before
+ * the upstream connection has been established. To prevent other filters disabling this, StateType
+ * 'ReadOnly' should be used. LifeSpan of this object is Connection. No sharing with the upstream
+ * should be needed.
+ *
+ * Note that this state needs to be set in filter's initializeReadFilterCallbacks() as tcp_proxy
+ * checks for this in its initializeReadFilterCallbacks() that is called when the filter chain is
+ * created, but after all the other filters as tcp_proxy is the last filter in the chain.
+ */
+class ProxyReceiveBeforeConnect : public StreamInfo::FilterState::Object {
```

Comment:
I don't think we need a new class for this; you can use `StreamInfo::BoolAccessor`. Also, the definition of the key belongs in `source/common/tcp_proxy/tcp_proxy.h`.

Related example that was recently added: `constexpr absl::string_view DisableTunnelingFilterStateKey = "envoy.tcp_proxy.disable_tunneling";`.

Reply:
Thanks for the pointers, changed!